pySDC/implementations/problem_classes/GenericGusto.py

```diff
@@ -0,0 +1,264 @@
+                map_if_true=drop,
+            )
+            mass_form = self.residual.label_map(
+                lambda t: t.has_label(time_derivative),
+                map_if_true=replace_subject(self.x_out, old_idx=self.idx),
+                map_if_false=drop,
+            )
+
+            problem = fd.NonlinearVariationalProblem((mass_form + residual).form, self.x_out, bcs=self.bcs)
+            solver_name = self.field_name + self.__class__.__name__
+            self.solvers['eval_rhs'] = fd.NonlinearVariationalSolver(
+                problem, solver_parameters=self.solver_parameters, options_prefix=solver_name
+            )
+            self.work_counters['solver_setup']()
+
+        self.solvers['eval_rhs'].solve()
+        self.work_counters['rhs']()
+
+        return self.dtype_f(self.x_out)
+
+    def solve_system(self, rhs, factor, u0, *args):
+        self.x_out.assign(u0.functionspace)  # set initial guess
+        self._u.assign(rhs.functionspace)
+
+        if factor not in self.solvers.keys():
+            if len(self.solvers) >= self.LHS_cache_size + self.rhs_n_labels:
+                self.solvers.pop(
+                    [me for me in self.solvers.keys() if type(me) in [float, int, np.float64, np.float32]][0]
+                )
+
+            # setup left hand side (M - factor*f)(u)
+            # put in output variable
+            residual = self.residual.label_map(all_terms, map_if_true=replace_subject(self.x_out, old_idx=self.idx))
+            # multiply f by factor
+            residual = residual.label_map(
+                lambda t: t.has_label(time_derivative), map_if_false=lambda t: fd.Constant(factor) * t
+            )
+
+            # subtract right hand side
+            mass_form = self.residual.label_map(lambda t: t.has_label(time_derivative), map_if_false=drop)
+            residual -= mass_form.label_map(all_terms, map_if_true=replace_subject(self._u, old_idx=self.idx))
+
+            # construct solver
+            problem = fd.NonlinearVariationalProblem(residual.form, self.x_out, bcs=self.bcs)
+            solver_name = f'{self.field_name}-{self.__class__.__name__}-{factor}'
+            self.solvers[factor] = fd.NonlinearVariationalSolver(
+                problem, solver_parameters=self.solver_parameters, options_prefix=solver_name
+            )
+            self.work_counters['solver_setup']()
+
+        try:
+            self.solvers[factor].solve()
+        except fd.exceptions.ConvergenceError as error:
+            if self.stop_at_divergence:
+                raise error
+            else:
+                self.logger.debug(error)
+
+        self.work_counters['ksp'].niter += self.solvers[factor].snes.getLinearSolveIterations()
+        self.work_counters['solver']()
+        return self.dtype_u(self.x_out)
+
+
+class GenericGustoImex(GenericGusto):
+    dtype_f = IMEX_firedrake_mesh
+    rhs_n_labels = 2
+
+    def evaluate_individual_term(self, u, label):
```

Comment:
slightly picky... but `label` might apply to more than one term so not sure "individual" is the right word